soql-standalone-parser/src/test/scala/com/socrata/soql/parsing/ParserTest.scala

```diff
@@ -231,29 +231,42 @@ class ParserTest extends WordSpec with MustMatchers {
 
     "window function over partition order round trip" in {
       val x = parseFull("select row_number() over(partition by x, y order by m, n)")
-      x.selection.expressions.head.expression.toString must be ("row_number() OVER ( PARTITION BY `x`,`y` ORDER BY `m` NULL LAST, `n` NULL LAST)")
+      x.selection.expressions.head.expression.toString must be ("row_number() OVER ( PARTITION BY `x`, `y` ORDER BY `m` ASC NULL LAST, `n` ASC NULL LAST)")
     }
 
     "window function over partition order desc round trip" in {
       val x = parseFull("select row_number() over(partition by x, y order by m desc null last, n)")
-      x.selection.expressions.head.expression.toString must be ("row_number() OVER ( PARTITION BY `x`,`y` ORDER BY `m` DESC NULL LAST, `n` NULL LAST)")
+      x.selection.expressions.head.expression.toString must be ("row_number() OVER ( PARTITION BY `x`, `y` ORDER BY `m` DESC NULL LAST, `n` ASC NULL LAST)")
     }
 
     "window function over partition round trip" in {
       val x = parseFull("select avg(x) over(partition by x, y)")
-      x.selection.expressions.head.expression.toString must be ("avg(`x`) OVER ( PARTITION BY `x`,`y`)")
+      x.selection.expressions.head.expression.toString must be ("avg(`x`) OVER ( PARTITION BY `x`, `y`)")
     }
 
     "window function over order round trip" in {
       val x = parseFull("select avg(x) over(order by m, n)")
-      x.selection.expressions.head.expression.toString must be ("avg(`x`) OVER ( ORDER BY `m` NULL LAST, `n` NULL LAST)")
+      x.selection.expressions.head.expression.toString must be ("avg(`x`) OVER ( ORDER BY `m` ASC NULL LAST, `n` ASC NULL LAST)")
     }
 
     "window function empty over round trip" in {
       val x = parseFull("select avg(x) over()")
       x.selection.expressions.head.expression.toString must be ("avg(`x`) OVER ()")
     }
 
+    "window function over partition frame" in {
+      val x = parseFull("select avg(x) over(order by m range 123 PRECEDING)")
+      x.selection.expressions.head.expression.toString must be ("avg(`x`) OVER ( ORDER BY `m` ASC NULL LAST RANGE 123 PRECEDING)")
+    }
+
+    "window frame clause should start with rows or range, not row" in {
+      expectFailure("Expression expected", "select avg(x) over(order by m row 123 PRECEDING)")
+    }
+
+    "order by nulls last should be null last w/o s" in {
```

Comment:
not sure I get this? isn't "nulls last" the correct syntax?

Reply:
sql/postgresql uses nulls.  unfortunately, soql uses null without s.

Reply:
aaaaaaaaaaaaaaaaa